cms/cache/placeholder.py

```diff
@@ -1,67 +1,183 @@
+        # of all its plugins during the rendering process anyway.
+        vary_on_list = placeholder.get_vary_cache_on(request)
+        duration = placeholder.get_cache_expiration(request, now())
+        # Update the main placeholder cache version
+        _set_placeholder_cache_version(
+            placeholder, lang, site_id, version, vary_on_list, duration)
+
+    sub_key_list = []
+    for key in vary_on_list:
+        value = request.META.get(get_header_name(key)) or '_'
+        sub_key_list.append(key + ':' + value)
+
+    cache_key = main_key
+    if sub_key_list:
+        cache_key += '|' + '|'.join(sub_key_list)
+
+    if len(cache_key) > 250:
+        cache_key = '{prefix}|{hash}'.format(
+            prefix=prefix,
+            hash=hashlib.md5(cache_key.encode('utf-8')).hexdigest(),
+        )
 
-def get_placeholder_page_cache(page_lookup, lang, site_id, placeholder_name):
+    return cache_key
+
+
+def set_placeholder_cache(placeholder, lang, site_id, content, request):
+    """
+    Sets the (correct) placeholder cache with the rendered placeholder.
+    """
     from django.core.cache import cache
 
-    return cache.get(_placeholder_page_cache_key(page_lookup, lang, site_id, placeholder_name),
-                     version=_get_cache_version())
+    key = _get_placeholder_cache_key(placeholder, lang, site_id, request)
+
+    duration = min(
+      get_cms_setting('CACHE_DURATIONS')['content'],
+      placeholder.get_cache_expiration(request, now())
+    )
+    cache.set(key, content, duration)
+    # "touch" the cache-version, so that it stays as fresh as this content.
+    version, vary_on_list = _get_placeholder_cache_version(placeholder, lang, site_id)
+    _set_placeholder_cache_version(
+        placeholder, lang, site_id, version, vary_on_list, duration=duration)
 
 
-def set_placeholder_page_cache(page_lookup, lang, site_id, placeholder_name, content):
+def get_placeholder_cache(placeholder, lang, site_id, request):
+    """
+    Returns the placeholder from cache respecting the placeholder's
+    VARY headers.
+    """
     from django.core.cache import cache
 
-    cache.set(_placeholder_page_cache_key(page_lookup, lang, site_id, placeholder_name),
-              content,
-              get_cms_setting('CACHE_DURATIONS')['content'], version=_get_cache_version())
-    _set_cache_version(_get_cache_version())
+    key = _get_placeholder_cache_key(placeholder, lang, site_id, request, soft=True)
+    content = cache.get(key)
+    return content
+
+
+def clear_placeholder_cache(placeholder, lang, site_id):
+    """
+    Invalidates all existing cache entries for (placeholder x lang x site_id).
+    We don't need to re-store the vary_on_list, because the cache is now
+    effectively empty.
+    """
+
+    current_version, _ = _get_placeholder_cache_version(placeholder, lang, site_id)
+    version = int(time.time() * 1000000)
+    if current_version >= version:
```

Comment:
I know is just in case but this is a relatively "impossible" scenario, don't think is necessary to have this extra code.